openhcl/underhill_core/src/emuplat/tpm.rs

```diff
@@ -2,109 +2,123 @@
-        }
+    pub fn new(tee_call: Box<dyn tee_call::TeeCall>) -> Self {
+        Self { tee_call }
     }
 }
 
 impl GetAttestationReport for TpmGetAttestationReportHelper {
-    fn get_attestation_report(
+    fn get_report(
         &self,
-        ak_pub_modulus: &[u8],
-        ak_pub_exponent: &[u8],
-        ek_pub_modulus: &[u8],
-        ek_pub_exponent: &[u8],
-        guest_input: &[u8],
+        report_data: &[u8; tee_call::REPORT_DATA_SIZE],
     ) -> Result<Vec<u8>, Box<dyn std::error::Error + Send + Sync>> {
-        let igvm_attest_request_helper =
-            underhill_attestation::IgvmAttestRequestHelper::prepare_ak_cert_request(
-                self.tee_call.tee_type(),
-                ak_pub_exponent,
-                ak_pub_modulus,
-                ek_pub_exponent,
-                ek_pub_modulus,
-                &self.attestation_vm_config,
-                guest_input,
-            );
-
         let result = self
             .tee_call
-            .get_attestation_report(&igvm_attest_request_helper.runtime_claims_hash)
-            .map_err(TpmGetAttestationReportError::GetAttestationReport)?;
+            .get_attestation_report(report_data)
+            .map_err(TpmAttestationError::GetAttestationReport)?;
 
-        let request = igvm_attest_request_helper
-            .create_request(&result.report)
-            .map_err(TpmGetAttestationReportError::CreateIgvmAttestAkCertRequest)?;
-
-        // Treat the request as the attestation report which will be exposed to the guest
-        // (via nv index).
-        Ok(request)
+        Ok(result.report)
     }
 }
 
 /// An implementation of [`RequestAkCert`].
 #[derive(Clone)]
 pub struct TpmRequestAkCertHelper {
     get_client: GuestEmulationTransportClient,
+    attestation_type: AttestationType,
+    attestation_vm_config: AttestationVmConfig,
     attestation_agent_data: Option<Vec<u8>>,
 }
 
 impl TpmRequestAkCertHelper {
     pub fn new(
         get_client: GuestEmulationTransportClient,
+        attestation_type: AttestationType,
+        attestation_vm_config: AttestationVmConfig,
         attestation_agent_data: Option<Vec<u8>>,
     ) -> Self {
         Self {
             get_client,
+            attestation_type,
+            attestation_vm_config,
             attestation_agent_data,
         }
     }
 }
 
 #[async_trait::async_trait]
 impl RequestAkCert for TpmRequestAkCertHelper {
-    async fn request_ak_cert(
+    fn create_ak_cert_request(
         &self,
-        attestation_report: Option<Vec<u8>>,
-    ) -> Result<Vec<u8>, Box<dyn std::error::Error + Send + Sync + 'static>> {
-        // Attestation report should be `Some` for CVM.
-        // TODO TVM: Allow attestation report in `None` when shared memory is supported
-        // (required by GET callout).
-        let Some(attestation_report) = attestation_report else {
-            return Ok(vec![]);
+        get_attestation_report: Option<&dyn GetAttestationReport>,
```

Comment:
why does this need to take a trait object + invoke it directly? can't you just pass a `attestation_report: Option<Vec<u8>>` instead?

Reply:
oh. I see. because it uses `ak_cert_request_helper.runtime_claims_hash`...

Reply:
if the TPM isn't using the `GetAttestationReport` trait at all, should it be removed from the TPM emulator's public interface, and have all functionality absorbed into `TpmRequestAkCertHelper`?

otherwise, you're just passing a trait object to the TPM emulator whose only purpose is to pass to another related trait object you're passing to the TPM emulator?